### DISCUSSION TITLE
Adafruit Metro ESP32-S2 and Adafruit Metro ESP32-M4 Airlift Lite with 4" TFT Touch Screen SPI (ILI9486 - 480x320 WaveShare)

### DISCUSSION THREAD

**Discussion by pierre-49** [Bodmer/TFT_eSPI#2351] [Compatible displays and setup files]
I made the settings in "Setup_Select.h" using the library "TFT_eSPI-master" and Arduino IDE 2.0.3.
The file is attached in PDF format.
Basically, only the arrangement of the pins relating to the display has been added to adapt them to the board by referring to the number of pins shown on the board itself (see attached figures - board and dfisplay).
Compiling some examples did not give rise to errors. Not even loading into the tab. But nothing works.
I'm not sure about the pin assignment in "Setup_Select.h".
This is because it is not clear whether the pins have the numbers referring to the board in question or referring to the Arduino UNO R3 pinout.
Any solution? Thanks.

![Adafruit Metro ESP32-S2 pinout](https://user-images.githubusercontent.com/48489090/214335134-6c905ddc-7299-4316-b054-9d784c6a22a1.png)
![Note e Pinout TFT vs Arduino 1 R3](https://user-images.githubusercontent.com/48489090/214334642-e0280ad0-8681-4308-a453-d07d036c8fea.png)

**Comment by pierre-49** [comment]
The lines in question are these:

// For ESP32-S2 Metro board (only tested with ILI9486 display)
// The hardware SPI can be mapped to any pins

#define TFT_MISO 21
#define TFT_MOSI 16
#define TFT_SCLK 42
#define TFT_CS 15 // Chip select control pin
#define TFT_DC 12 // Data Command control pin
#define TFT_RST 13 // Reset pin (could connect to RST pin)
#define TFT_BL 14 // LED backlight
#define TOUCH_CS 9 // Chip select pin (T_CS) of touch screen

[User_Setup.pdf](https://github.com/Bodmer/TFT_eSPI/files/10491560/User_Setup.pdf)

**Comment by Bodmer** [comment]
The GPIO pin numbers should be used for the setup file and I see you have done this.

There are a couple of difficulties with this board combination however with the correct allocations it should work (see below). Also note that the TFT shield is fitted with a set of chips that create a serial to parallel conversion. This interface arrangement is used in RPi type displays. Assuming that this interface circuit does what I expect then adding the following line to the setup file will invoke the correct interface code in the TFT_eSPI library:

#define RPI_DISPLAY_TYPE

With that interface design you may need to limit the SPI speed, speeds in the range 10MHz to 20MHz should work. You may find you can go higher in speed.

The Arduino board package for the ESP32-S2 does not map the SPI port to GPIO pins 21, 16 and 42 so that will not work. (This is an incompatibility in the UNO format of that Metro board.  Instead you must use the signals routed through the 6 pin ICSP port. Then use those GPIO numbers for the SPI pins. The TFT has a DIP switch to link the SPI signals to the ISCP port, it looks like these are in the right prosition but you will need to check the boards documentation.  It looks like this would be correct for the TFT_eSPI setup file allocation to the SPI pins:

#define TFT_MISO 37
#define TFT_MOSI 35
#define TFT_SCLK 36

**Reply by pierre-49** [reply]
Before anything else, many thanks for the clarification.
My fault. If I had paid more attention to the card's Datasheet, perhaps I could have done better.
From various checks, using the simple Arduino IDE, and specifically the:
Read_User_Setup         ---> works;
Test_Touch_Controller  ---> works;
Free_Font_Demo          ---> works but the BackLight is not driven and then the screen is black.
With reference to this last item, the strange thing is that with the ignition switch (on the card) in the ON position (see image), the command of the BL (GPIO14 - PIN 14) is always at 0.2V; when switching the switch to the OFF position, the voltage rises to 0.8V.
If you place the board with switch in the ON position and then switch it to OFF, for a few seconds (until ESP, which is no longer powered, still works), the screen lights up and you can see that the fonts are there present.
So it seems that for some reason the GPIO14 is not initialized and/or not commanded.

Thanks in advance for the reply.
![Adafruit Metro ESP32-S2 Pag9](https://user-images.githubusercontent.com/48489090/214530385-7e865e39-20b2-4379-b78f-a2e0bcf0470e.jpg)

**Reply by pierre-49** [reply]
I solved! The enablement of this define in "User_Setup.h" was missing:
#define TFT_BACKLIGHT_ON HIGH

Now, at least the "Free_Font_Demo" test works perfectly!

Thanks anyway.

--> After having tested many sketches, I will release the setups for this type of TFT.

**Comment by pierre-49** [comment]
Just today I got the Adafruit Metro ESP32-M4 AirLift Lite board. The pinout is in the figure.
With the older Adafruit Metro ESP32-S2 as an example I set up a new setup file. For the Metro ESP32-S2 it works perfectly (attached "Setup11a_RPi_touch_ILI9486.h" in pdf format), for the Metro ESP32-M4, with the changed assignments it doesn't work at all.
In the file the assignment was made with the pin number present on the card and evidently it is not correct.
Obviously this was a first attempt doomed to failure.
It is clear that there is not only a pin assignment problem, but since the ESP32-S2 Metro board has a single processor and the output pins on the connectors are connected to it, everything works.
In the ESP32-M4 Metro board there are two processors and the pins on the connectors are connected to the SAM51 processor.
And therefore they are not reachable by the current library software.
Any tips on how and where to operate? I could implement it myself.

Thank you.
![Adafruit Metro M4 Express AirLift WiFi Pinout](https://user-images.githubusercontent.com/48489090/214634513-25288bae-488c-499e-bd9d-704a4b02203d.png)
[Setup11a_RPi_touch_ILI9486.pdf](https://github.com/Bodmer/TFT_eSPI/files/10501763/Setup11a_RPi_touch_ILI9486.pdf)
[Adafruit Metro M4 Express AirLift WiFi Schematic.pdf](https://github.com/Bodmer/TFT_eSPI/files/10501764/Adafruit.Metro.M4.Express.AirLift.WiFi.Schematic.pdf)

**Reply by Bodmer** [reply]
That RPi interface screen type is not supported by TFT_eSPI when used with the SAMD51 M4 processor.

**Comment by pierre-49** [comment]
For the same 480x320 WaveShare (SPI) display and an AZ-Delivery ESP32 Dev KitC V2 card, what is the suitable setup and what pin assignment, which on the card have names and not numbers?

Thanks in advance.

![ESP_-_32_NodeMCU_Developmentboard_Pinout_Diagram](https://user-images.githubusercontent.com/48489090/218712628-a304bc2c-04f2-4c81-a611-3b97eb8ccafc.jpg)
